nntrainer/include/properties.h

```diff
@@ -0,0 +1,363 @@
+// SPDX-License-Identifier: Apache-2.0-only
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file properties.h
+ * @date 03 Aug 2020
+ * @brief Properties handler for NNtrainer.
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+
+#ifndef __NNTRAINER_PROPERTIES_H__
+#define __NNTRAINER_PROPERTIES_H__
+#ifdef __cplusplus
+
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+template <typename T> struct prop_traits {
+  static constexpr bool is_property = false; /**< check if this is property */
+
+  typedef bool value_type; /**< actual value type of given property */
+
+  /**
+   * @brief validation logic for type T
+   * @param[in] value Target to have registration
+   */
+  static bool is_valid(value_type value) {
+    throw exception::not_supported("no trait type can be validated");
+  }
+
+  /**
+   * @brief get string key of current property
+   */
+  static std::string getKey() {
+    throw exception::not_supported("no trait type have a key");
+  }
+
+  /**
+   * @brief get default value of current type
+   */
+  static value_type getDefault() {
+    throw exception::not_supported("no trait type does not have default");
+  }
+
+  /**
+   * @brief serialization for type T
+   * @param[in] target Target to be converted to string
+   */
+  static std::string serialize(value_type target) {
+    throw exception::not_supported("Only property type can be serialized");
+  }
+
+  /**
+   * @brief deserialization for type T
+   * @param[in] str string to be converted to actual type T
+   * @throw exception::not_supported if T is not property
+   * @throw std::invalid_argument if @a str is not deserializable.
+   */
+  static value_type deserialize(std::string &str) {
+    throw exception::not_supported("Only property type can be deserialized");
+  }
+};
+
+/*****************************************************************************
+ *  Property registration part.                                              *
+ *  If you want to add a property, only this part need change.               *
+ *  You only have to add property struct and prop_traits.                    *
+ *                                                                           *
+ *  It is recommened to use unique key for each types.                       *
+ *  However, it is allowed to use duplicated key                             *
+ *  When those properties are mutually exclusive                             *
+ *****************************************************************************/
+
+/**< color is here for demonstration purpose. */
+
+enum class Color {
```

Comment:
I don't think we need to demonstrate in header. comment out this codes or have to add this in test codes.